Adds the gas condenser

TemporalOroboros opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
- Adds a machine that can be used to condense gases to reagents.
- Changes how certain reagents create gas to prevent infinite gas production.
  - Specifically removes the generating gas on splash effect and moves the gas production over to boiling the reagent.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
- More interdepartmental interaction
- Less dependence on the chem dispenser

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Adds the gas condenser. Now you can bug atmospherics if the miners are refusing to give you plasma to grind!
balance: Certain reagents no longer generate gas when splashed. Instead, they can be boiled in any open top container.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Yes the condenser sprite is just a modified freezer. Sue me, I can't sprite for my life.

TemporalOroboros committed:
Adds the gas condenser and boiling reagents

- Adds the Gas Condenser. A machine that lets you convert gases into reagents!
- Adds the boilable component and attaches it to several reagents.

TemporalOroboros committed:
Updates condensation rates to use delta_time. Adds enthalpies of vaporization

- Makes condensation rate use delta_time
- Gives gases enthalpies of vaporization
- Fixes nulls in condensation heats being interpreted as 1

LemonInTheDark commented:
you appear to have changed freezer sprites, 🗞️

TemporalOroboros commented:
@LemonInTheDark I have no idea why freezer sprites changed. I'll fix it.

TemporalOroboros committed:
Reverts thermomachine icon change and fixes white pixel

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_types.dm:
```diff
@@ -40,32 +68,99 @@ GLOBAL_LIST_INIT(nonreactive_gases, typecacheof(list(/datum/gas/oxygen, /datum/g
+	/// Relative rarity compared to other gases, used when setting up the reactions list.
+	var/rarity = 0
+
+	// Condensation:
+	/// The maximum rate of condensation of this gas (as a function of total moles). If this is falsey the gas will not condense
+	var/cond_rate
+	/// The maximum temperature this gas will condense at
+	var/cond_temp_max
+	/// The temperature at and below which the gas will condense at it's maximum rate
+	var/cond_temp_min
+	/// The typepath of the reagent this gas condenses into. If this is null the reagents will not produce a gas when condensing
+	var/datum/reagent/cond_type
+	/// The amount of thermal energy released by one mole of this gas condensing
+	var/cond_heat = 0
+	/// A reference to the condensation event [CALLBACK] If this is truthy on init it will be repaced with the appropriate callback. Leave this as null
+	var/datum/callback/cond_event = null
```
Why is this being done on gas_types? And why are you treating them as singletons rather then templates

TemporalOroboros replied:
Because I want different gases to condense at different temperatures and rates. The singleton bit is just so it can generate the callback for condensation effects.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_types.dm:
```diff
@@ -40,32 +68,99 @@ GLOBAL_LIST_INIT(nonreactive_gases, typecacheof(list(/datum/gas/oxygen, /datum/g
 \*||||||||||||||||||||||||||||||||||||||||*/
 
 /datum/gas
+	/// A reference to the cached gas metadata list
+	var/list/meta_gas_list
```
What the fuck

TemporalOroboros replied:
Easy reference for debugging exactly where you'd expect it to be without looking at the code. I might as well as long as I'm caching them.

LemonInTheDark replied:
why cache them in the first place? I don't understand

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_types.dm:
```diff
@@ -19,8 +19,36 @@ GLOBAL_LIST_INIT(nonreactive_gases, typecacheof(list(/datum/gas/oxygen, /datum/g
+
 		.[gas_path] = gas_info
 
+	if(SSair)
+		SSair.gas_metadata = .
```
Why

TemporalOroboros replied:
It's easier to access when debugging

LemonInTheDark replied:
cause of the pain of sorting through GLOB vs SSair?

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_types.dm:
```diff
@@ -123,6 +233,19 @@ GLOBAL_LIST_INIT(nonreactive_gases, typecacheof(list(/datum/gas/oxygen, /datum/g
+	cond_event = TRUE
+	cond_heat = 1390
+
+/datum/gas/tritium/on_condense(datum/gas_mixture/holder, cond_moles, cond_temp, atom/location, datum/reagents/target, list/cond_rates)
+	. = ..()
+	if(cond_moles < TRITIUM_MINIMUM_RADIATION_ENERGY)
+		return
+	if(prob(50))
+		radiation_pulse(location, cond_moles / TRITIUM_MINIMUM_RADIATION_ENERGY)
```
Gas types are not objects, they're templates. Why is this being done here rather then in some condensation specific bit of code?

TemporalOroboros replied:
Because if I chucked it into condensation code it would be more difficult to expand on later.

LemonInTheDark reviewed (changes requested):
I have not reviewed the component itself yet, I have issues with turning gas datums into singletons rather then templates, and want to know why all this behavior is being handled/defined in this bit of code.

TemporalOroboros commented on code/__DEFINES/reagents.dm:
```diff
@@ -2,6 +2,25 @@
+/// The boiling point of water.
+#define BP_WATER			(T0C+100)
+/// The boiling point of nitrogen.
+#define BP_NITROGEN			(300+10)
+/// The boiling point of oxygen.
+#define BP_OXYGEN			(300+10)
```
Because if I make it the actual boiling point of nitrogen/oxygen it will boil the instant it comes out of the chem dispenser and I don't want to drive chemists up a wall. This can be changed once the primary way of acquiring oxygen/nitrogen is not dispensing it in liquid form into an open-topped container.

LemonInTheDark replied:
No like why is it number + number rather then either 310 or TOC + 10

TemporalOroboros replied:
My brain was still running on the same logic that made water's boiling point T0C + 100. I'll fix it when I get back to a computer.

TemporalOroboros replied:
Should be done now.

TemporalOroboros commented on code/controllers/subsystem/air.dm:
```diff
@@ -26,6 +26,11 @@ SUBSYSTEM_DEF(air)
 	var/list/pipe_init_dirs_cache = list()
 
 	//atmos singletons
+	/// Lazy list of the gas meta info lists. Literally just a reference to GLOB.meta_gas_list
+	var/list/gas_metadata
+	/// Lazy list of gas datum singletons. Currently used exclusively to generate condensation event singletons
+	var/list/gas_singletons
```
I need them to hold a proc

TemporalOroboros commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -536,3 +537,78 @@ get_true_breath_pressure(pp) --> gas_pp = pp/breath_pp*total_moles()
+		if(cached_temp >= cond_max_temp)
+			continue
+
+		var/cond_rate = clamp(LERP(0, cond_max_rate, (cond_max_temp - cached_temp) / (cond_max_temp - cond_min_temp)), 0, 1)
+		cond_rate = clamp(cached_gases[gas_id][MOLES] * DT_PROB_RATE(cond_rate, delta_time), 0, cached_gases[gas_id][MOLES])
```
The condensation rate starts out as a ratio of the gas's moles that condense every second. This is calculated from the temperature of the gas and its minimum/maximum condensation temperatures. Lower is better and it caps out at the min... Actually I should really clamp that bit inside of the `LERP()` to prevent temperatures below `cond_min_temp` further increasing condensation rate. Anyway, this is then scaled using `DT_PROB_RATE()` by delta_time since it accurately reflects the scaling. This is then multiplied by the moles of the gas to get the final condensation rate.

TemporalOroboros commented on code/modules/atmospherics/gasmixtures/gas_types.dm:
```diff
@@ -19,8 +19,36 @@ GLOBAL_LIST_INIT(nonreactive_gases, typecacheof(list(/datum/gas/oxygen, /datum/g
+	if(SSair)
+		SSair.gas_metadata = .
+
+/proc/init_gas_singletons()
+	. = list()
+	for(var/gas_path in subtypesof(/datum/gas))
+		var/datum/gas/gas_singleton = new gas_path
+		.[gas_path] = gas_singleton
+
+		if(GLOB.meta_gas_info) // We do this twice to make sure it gets in regardless of which order these are instantiated
+			gas_singleton.meta_gas_list = GLOB.meta_gas_info[gas_path]
+			GLOB.meta_gas_info[gas_path][META_GAS_COND_EVENT] = gas_singleton.cond_event
+
+	if(SSair)
+		SSair.gas_singletons = .
```
It's entirely to generate the callback used for condensation effects.

LemonInTheDark replied:
hhhhhhhhhhhhhhhhhhhhhhhhhhhh

TemporalOroboros committed:
Update code/modules/atmospherics/gasmixtures/gas_mixture.dm

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

TemporalOroboros commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -536,3 +537,78 @@ get_true_breath_pressure(pp) --> gas_pp = pp/breath_pp*total_moles()
+ * - delta_time: The amount of time the condensation is happening over
+ */
+/datum/gas_mixture/proc/condense_gas_to(atom/location, datum/reagents/reagents, delta_time)
+	if(delta_time <= 0)
+		return
```
Because, unfortunately, the delta_time passed by the condenser cannot be trusted to be positive. It basically exists in case some admeme vv's a condensers condensation efficiency into the negatives.

TemporalOroboros commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -147,6 +147,11 @@
+
+/datum/reagent/water/on_new(data)
+	. = ..()
+	AddElement(/datum/element/boilable, T0C+100, "h2o")
```
Probably.

TemporalOroboros commented on code/datums/elements/boilable.dm:
```diff
@@ -0,0 +1,43 @@
+	if(!location)
+		return NONE
+
+	location.atmos_spawn_air("[vapor_path]=[source.volume * REAGENT_MOLE_DENSITY];TEMP=[_temp]")
```
It depends on how much smoke and/or foam you used to spread the plasma. If you used enough it's less, if you used too little it's more.

TemporalOroboros replied:
Bassically it's a sort of middle ground for gas production since you currently can get varying yields depending on how much smoke you use to apply the reagents. I am removing the gas production on smoke to prevent infinite gas production with a smoke machine.

TemporalOroboros committed:
Used BP_WATER and clamps condensation temp

TemporalOroboros committed:
Merge branch 'gas-condenser' of https://github.com/TemporalOroboros/tgstation into gas-condenser

TemporalOroboros commented on code/__DEFINES/reagents.dm:
```diff
@@ -2,6 +2,25 @@
+/// Moles per reagent unit. Used to convert between reagent units and gases.
+#define REAGENT_MOLE_DENSITY	10
+/// The boiling point of water.
+#define BP_WATER			(T0C+100)
```
Should be fixed

Angustmeta commented:
Please explain further for muh wiki. 

1.  Will smoke now only release gases if the grenade heats the (liquid plasma etc) over their boiling point in the same reaction as the smoke? You mentioned smoke/foam in a comment above but I didn't quite get it. 
2. Will smoke still multiply gases like plasma to each tile? How does smoke amount affect total moles released per unit? 
4. Will foam only spread gases if the same reaction heated them above boiling point? 
5. How does foam amounts affect total moles released per unit?
6. Are foam or smoke changed in any way except the way they release gases?

TemporalOroboros commented:
@Angustmeta I'm sorry if the wording was a bit confusing. I've completely removed the ability for nitrogen, oxygen, plasma, and similar reagents to produce gas when splashed, sprayed, smoked, or foamed. All of the given reagents now only produce gas when they are heated above their boiling points in an open topped container. The reason this was done was to prevent the ability to infinitely produce gas by plugging a scrubber into a condenser into a smoke machine since producing gas through smoke scales with the size of the smoke cloud and there 
for can get you more gas than you used to condense the reagents in the smoke regardless of the relevant reagent vaporization/gas condensation ratios. It may be re-implemented in the future, I have some ideas, but those implementations will just produce the gas when the smoke cloud starts and the smoke and foam won't be able to spread the reagents.

Angustmeta commented:
Thanks for clarifying.

Losenis commented:
By open-topped, do you mean acclimators and pyrosium reaction chambers will not have them turn into gas?

TemporalOroboros commented:
@Losenis I am considering anything you can draw reagents from by smacking it with a beaker to be open topped. So chemical acclimators, reaction chambers, syringes, slime cores, etc will not allow reagents inside them to boil.

Cyprex commented:
> The reason this was done was to prevent the ability to infinitely produce gas by plugging a scrubber into a condenser into a smoke machine since producing gas through smoke scales with the size of the smoke cloud and therefor can get you more gas than you used to condense the reagents in the smoke regardless of the relevant reagent vaporization. The reason this was done was to prevent the ability to infinitely produce gas by plugging a scrubber into a condenser into a smoke machine since producing gas through smoke scales with the size of the smoke cloud and therefor can get you more gas than you used to condense the reagents in the smoke regardless of the relevant reagent vaporization/gas condensation ratios./gas condensation ratios.

I think this an interesting mechanic and don't think it's worth removing for this feature.

Losenis commented:
Would that mean that an acclimator heating reagents past boiling point to then move them to an output gate will have them boil?

TemporalOroboros commented:
@Losenis Yes

TemporalOroboros committed:
Simplifies oxygen and nitrogen boiling points

- @LemonInTheDark

TemporalOroboros committed:
Merge branch 'master' into gas-condenser

TemporalOroboros committed:
Fixes a couple bugs I found it testing + boiling conditions

- CO2 will no longer div/0 when you try to condense it. Code is guarded against potential future issues of the same type.
- Relevant reagents now actually receive the boilable component. This may have broken some others; more testing it required.
- Changed back to using typepaths for the boilable component
- Expanded definition of open topped containers to include anything you can fill with a beaker and anything you can dunk things into.

TemporalOroboros committed:
Makes cryobeakers able to prevent reagents from boiling.

- If they can keep high explosives from _exploding_ they can keep water from boiling.

TemporalOroboros committed:
Guards every coder's least favorite spirit against getting passed null data

- Fruit wine won't runtime if it gets instantiated without input data.

TemporalOroboros committed:
Fixes nitrous oxide boiling out of the container during unit tests

- Adjusts the temperature of the nitrous oxide added to the beaker to be below the boiling point.

TemporalOroboros commented:
@Cyprex I'll try to figure something out that doesn't result it egregious gas duping.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_types.dm:
```diff
@@ -40,32 +68,99 @@ GLOBAL_LIST_INIT(nonreactive_gases, typecacheof(list(/datum/gas/oxygen, /datum/g
 	specific_heat = 30
 	name = "Carbon Dioxide"
 	rarity = 700
+	cond_rate = 0.05
+	cond_temp_max = BP_CARBON_DIOXIDE
+	cond_temp_min = BP_CARBON_DIOXIDE
+	cond_type = /datum/reagent/carbondioxide
+	cond_heat = 25230
```
Why did you use the numbers you did here/in the other gastypes?

TemporalOroboros replied:
The condensation rate was pretty much arbitrary. The max condensation temp is just the temperature it boils at. The minimum condensation temp was again, pretty much arbitrary. I think I chose it to be equal to the condensation temp because CO2 turns into dry ice? I don't remember. I just googled the IRL enthalpy of vaporization for the condensation heat.

LemonInTheDark commented on code/modules/atmospherics/machinery/components/unary_devices/condenser.dm:
```diff
@@ -0,0 +1,268 @@
+	for(var/obj/item/stock_parts/matter_bin/bin in component_parts)
+		total_bin_rating += bin.get_part_rating()
+
+	condensation_rate_multiplier = total_bin_rating * CONDENSER_RATE_BIN_SCALE
```
can't this make infinite gas?

TemporalOroboros replied:
That's just a multiplier for the rate the gas condenses at. It doesn't change how much reagent you get out of the gas.

LemonInTheDark reviewed (commented):
More comments, I can't come up with an honest replacement for using gas types as singletons, as mad as it makes me.
Could you elaborate on why this is good for the game?

TemporalOroboros commented on code/controllers/subsystem/air.dm:
```diff
@@ -70,6 +75,12 @@ SUBSYSTEM_DEF(air)
 	gas_reactions = init_gas_reactions()
 	setup_turf_visuals()
+
+	if(GLOB.meta_gas_info)
+		gas_metadata = GLOB.meta_gas_info
+	if(GLOB.gas_singletons)
+		gas_singletons = GLOB.gas_singletons
```
I heard that the order the subsystems and GLOB initialize in is inconsistent. There's references to as much in reagent and material code at least. I didn't want to deal with rewriting how the gas list is built so I made it so the data should get filled out regardless of which order they get built in.

TemporalOroboros commented:
@LemonInTheDark 
https://github.com/tgstation/tgstation/pull/55585#discussion_r547154461

I honestly don't remember my exact reasoning. I think it was just for debugging convenience while I was making them singletons.

As for why this is good for the game... It makes chemistry slightly less dependent on the chemistry dispenser. There's the whole plan to axe the thing entirely after all. Other than that it slightly increases department interaction and, once the dispenser's gone, might make for a good way to actually get the gas reagent. I don't really see botany or ore producing hydrogen very often.

TemporalOroboros committed:
Merge branch 'master' into gas-condenser

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into gas-condenser

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into gas-condenser

TemporalOroboros committed:
Re-fixes reagents boiling. Makes everything use set_temperature., Etc.

- Fixes reagent boiling signal hook
- Makes everything that modifies reagent temp use `set_temperature()`
- Adjusts condenser overpressure/underpressure/overheat sound effects.

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Clean up some redundant GLOB gas list code. WARNING: May drive lemon up the wall

- Refactors how gas lists are created
- Adds VV support to gases
- Adds future support for making emissive gases

TemporalOroboros committed:
Fixes circuit board and design icon states

- Makes the condenser machine board use the engineering icon state
- Makes the condenser's initial icon state a valid state

TemporalOroboros committed:
Adjusts args on boilable Attach

TemporalOroboros committed:
Adjusts order of operations in reagent addition

- Fixes the reagent exposure unit test

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into upstream/master

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into gas-condenser

TemporalOroboros committed:
Merge branch 'master' into gas-condenser

LemonInTheDark commented:
Gonna give this another readover, I can't exactly pin down my issues with the old/current implementation, I think it mostly stems from putting so much bloat into gastype code, though I suppose this is about what it's for. I'll need to weigh the cost/benefit of the feature versus the code quality hit

TemporalOroboros committed:
Merge branch 'master' into gas-condenser

TemporalOroboros committed:
Merge branch 'master' into gas-condenser

TemporalOroboros committed:
Merge branch 'master' into gas-condenser

TemporalOroboros committed:
Merge branch 'master' into gas-condenser

TemporalOroboros committed:
Fixes div/0 error when the condenser spews foam

TemporalOroboros committed:
Merge branch 'master' into gas-condenser

TemporalOroboros committed:
Revert ".yarn copyright things?"

This reverts commit 1829cf7c56214653aa1a05f4847873b288b2780e.

TemporalOroboros committed:
.yarn copyright things?

stylemistake reviewed (changes requested):
Please unchange changes in `.yarn` directory.

LemonInTheDark commented:
While I think the idea itself is harmless, I do not think it is worth this amount of detail, especially in gas_types. It's too complex, and the readability hit is greater then the benefit of turning gases into reagents.
 
I'd love to see the the improvements littered throughout this pr in future, along with the boilable element, but the main thrust of condensation is just too complex for too little return. Honestly I'm kind of annoyed we store fusion power in the first place, I don't want to double the size of the meta gas list for one feature.
 
I'm sorry for the long delay on this. I hope it hasn't taken up too much more of your time.

TemporalOroboros committed:
Merge branch 'master' into gas-condenser

TemporalOroboros commented:
@LemonInTheDark Is there anything specific you want me to atomize out of this?
